Fixes #22254 - Add autocomplete component

danseethaler opened the pull request:
Add the downshift package to implement an autocomplete component.
Currently the component accepts `items`, `onInputUpdate`, `labelText`,
`actionText`, and `onSelect`.

http://projects.theforeman.org/issues/22254

danseethaler committed:
Fixes #22254 - Add autocomplete component

Add the downshift package to implement an autocomplete component.
Currently the component accepts items, onInputUpdate, labelText,
actionText, and onSelect.

http://projects.theforeman.org/issues/22254

amirfefer commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  componentDidMount() {
+    this.setState({ containerWidth: `${this.inputRef.offsetWidth}px` });
```
couldn't we use pure css styling, something like  `width: inherit` ?
Using state and ref only for setting element width looks a bit redundant, I tend to try make it stateless functional component, especially if we would have a redux wrapper component.

danseethaler replied:
Yeah good call on this. I wanted to have the dropdown match the width of the input box but I guess including the search button is fine and this will make the code simpler. Thanks!

amirfefer commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/AutoComplete.test.js:
```diff
@@ -0,0 +1,17 @@
+        onSelect={() => {}}
+      />);
+    expect(toJson(wrapper)).toMatchSnapshot();
+  });
```
Does it full coverage ?

danseethaler replied:
It does cover the whole component but no items are rendered. I struggled with writing tests for this component because most of what you would think to test is actually implemented by the [downshift](https://github.com/paypal/downshift) library. The only thing we're doing is implementing the actual html/css for the dropdown. The items state is managed by whatever invokes this component and all the inner workings are handled by downshift.

sharvit replied:
@danseethaler you can create a manual mock inside the `__mocks__` folder:
https://facebook.github.io/jest/docs/en/manual-mocks.html

danseethaler replied:
@sharvit thanks for the idea. That got me thinking and I ended up abstracting the rendering of items so I could test it independently. Now we've got full coverage on the component. Thanks!

waldenraines commented:
@danseethaler is this being used anywhere in foreman or katello so I can test it out or just the storybook?

danseethaler commented:
@waldenraines I have a branch in Katello I've started to use it - but I don't have a PR yet. https://github.com/danseethaler/katello/tree/search-box-21953

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import Downshift from 'downshift';
+import classnames from 'classnames';
+import { InputGroup } from 'react-bootstrap';
```
You should be able to import it from `patternfly-react`
https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Disabled=false&selectedKind=Forms&selectedStory=Input%20Groups&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,79 @@
+        </InputGroup>
+
+        {isOpen && items.length ? (
+          <div className="autocomplete-dropdown-container">
```
I think you can use `DropdownMenu` and `MenuItem` from `patternfly-react`, then you should be able to remove your `scss` file
https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Disabled=false&knob-Style=default&selectedKind=MenuItem&selectedStory=MenuItem&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/AutoComplete.stories.js:
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable no-alert */
+import React, { Component } from 'react';
+import { storiesOf } from '@storybook/react';
+import { Row, Col } from 'react-bootstrap';
```
Row and Col can be imported from `patternfly-react`
I think you can also import `Grid` and do `Grid.Row` `Grid.Col`

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,90 @@
+      <div>
+        {labelText ? <label {...getLabelProps()}>{labelText}</label> : null}
+        <InputGroup>
+          <input
```
You can build your form with `patternfly-react` components only:
https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Disabled=false&knob-Style=default&knob-Show%20Labels=false&knob-Show%20Help=true&knob-Show%20Loading=false&selectedKind=Forms&selectedStory=Supported%20Controls&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,90 @@
+export const renderItems = ({
+  items, highlightedIndex, selectedItem, getItemProps,
+}) => (
+  <div className="autocomplete-dropdown-container">
```
You can build this container with `patternfly-react`, i think you won't need to write any class or css:
[MenuItem](https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html?knob-Disabled=false&knob-Style=default&knob-Show%20Labels=false&knob-Show%20Help=true&knob-Show%20Loading=false&selectedKind=MenuItem&selectedStory=MenuItem&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,90 @@
+}) => (
+  <Downshift
+    onChange={onSelect}
+    itemToString={i => (i == null ? '' : String(i))}
```
Wouldn't it be easier to do `i ? String(i) : ''`

sharvit reviewed (changes requested):
Thanks @danseethaler - i have been playing with the storybook and it feels good!
I think you can write less code and based staff on `patternfly-react`.

waldenraines commented:
@danseethaler on the RH repos page example (at least) the dropdown is pushing the entire page down.  We need to make it an overlay over the page content instead.

<img width="1225" alt="screen shot 2018-01-19 at 11 41 57 am" src="https://user-images.githubusercontent.com/4116405/35161324-dc92dfe4-fd0d-11e7-9a49-82edcdcc9336.png">

danseethaler commented:
@waldenraines hm yeah that shouldn't be. I'll take another look at it.

@sharvit great recommendations! I'll take a look.

danseethaler commented:
Alright I think I've fixed all the issues that have been brought up. The `onClick` event in downshift stopped firing when I added the `MenuItem` component. It took a while but I finally figured out why and added an override [here](https://github.com/theforeman/foreman/pull/5169/files#diff-48abb5989a92452a1730b35af83f3cd6R29). Feel free to review again @waldenraines @sharvit

waldenraines commented:
@danseethaler does this work with multiple calls to autocomplete after the user starts typing.  I don't see any other popups after the initial autocomplete (even though the API is being called).

danseethaler commented:
@waldenraines yeah this component actually doesn't handle what items are displayed at all. It simply renders the items you give it and tells you when the search parameters update.

I'm sure my branch that has the search in katello needs some improvement. I'm hoping back into that branch right now to work on it some more.

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,97 @@
+      selectedItem,
+    }) => (
+      <div>
+        {labelText ? <label {...getLabelProps()}>{labelText}</label> : null}
```
Can be:
```js
 {labelText && <label {...getLabelProps()}>{labelText}</label>}
```

sharvit commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,97 @@
+        </InputGroup>
+
+        {isOpen && items.length
+          ? renderItems({
```
Can be:
```js
 {isOpen && items.length && renderItems({ ... })}
```

sharvit reviewed (changes requested):
Thanks @danseethaler, looks really good!

1. I notice an `console.error` when running the tests, any idea why? (haven't notice it at the storybook)
```bash
npm test -- autocomplete

> TheForemanDevDeps@1.18.0 test /home/asharvit/Projects/theforeman/foreman
> node node_modules/.bin/jest "autocomplete"

 PASS  webpack/assets/javascripts/react_app/components/common/AutoComplete/AutoComplete.test.js
  Autocomplete component
    ✓ should render input but not items on initial render (18ms)
    ✓ renderItems should render the items (10ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   2 passed, 2 total
Time:        1.749s
Ran all test suites matching "autocomplete".
  console.error node_modules/fbjs/lib/warning.js:33
    Warning: Each child in an array or iterator should have a unique "key" prop.
    
    Check the top-level render call using <DropdownMenu>. See https://fb.me/react-warning-keys for more information.
        in MenuItem
```

2. After i choose an item from the list, i can't clear the input and can't make changes that doesn't fit with the available options.
For example:
Choose item -> clear the input or change it to 'foo' -> the previous selection appearing back.
* Is this what we want from this component?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,113 @@
+  items, highlightedIndex, selectedItem, getItemProps,
+}) => (
+  <Dropdown.Menu
+    style={{
```
I would avoid inline style if its static?

ohadlevy commented on webpack/assets/javascripts/react_app/components/common/AutoComplete/index.js:
```diff
@@ -0,0 +1,113 @@
+      <MenuItem
+        {...getItemProps({
+          index,
+          item,
```
in the search auto completer we currently have, we also have a sub section for operators (and, or etc) do you think this component should support it or rather another high level component?

danseethaler replied:
This is a really good question. I think we do need to support these sub-sections which means we'll need to support a configuration being passed in rather than just an array of strings. I'll work on adding that in.

ohadlevy reviewed (commented):
overall looks good, a few inline comments + general question:

This is currently not used in foreman right? are you aiming to replace the search auto completer? I'm only asking as the search auto completer needs to handle a few more things like:
1. api calls  (or redux actions)
2. limit the amount of time you make an api calls if the user is typing fast
3. have a good  keyboard support (esc, up, down, space etc)
4. auto submit on mouse selection or similar.

As it stands now this component is fairly generic? maybe it should exists in pf-react instead?

danseethaler commented:
Thanks for the feedback @ohadlevy. This component does support keypress navigation/selection, triggers a callback when the submit/search button is pressed. The api calls are all outside of this component to enable using any api you like. You can see the example in the `.stories` file.

I think it's a good idea to at least bring it up to the patternfly community so I opened an issue https://github.com/patternfly/patternfly-react/issues/189. We'll see how the community responds.

danseethaler commented:
Also you can play with the storybook to get a sense of the component [here](https://rawgit.com/danseethaler/patternfly-react/autocomplete-component-storybook/index.html?selectedKind=AutoComplete&selectedStory=AutoComplete&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs).

danseethaler commented:
Going to close this as the component is progressing in the patternfly-react lib.
